docs/funnel.md

```diff
@@ -0,0 +1,228 @@
+# funnels [<sup><mark>experimental</mark></sup>](/docs/README.md#tag-notes)
+
+- Current status:	prototype
+- Effort remaining:	lots
+
+## Purpose
+
+TODO
+
+## Use cases
+
+TODO
+
+### Test Data
+
+Examples below are tested against the following data:
+
+```SQL ,non-transactional
+DROP TABLE IF EXISTS funnel_test;
+CREATE TABLE funnel_test(
+	"time" TIMESTAMPTZ,
+	"user" INTEGER,
+	"event" TEXT);
+INSERT INTO funnel_test VALUES
+	('2020-01-01 00:00:00+00', 1, 'LOGIN'),
+	('2020-01-02 00:00:00+00', 1, 'LOGIN'),
+	('2020-01-02 01:00:00+00', 1, 'SEARCH'),
+	('2020-01-02 03:00:00+00', 1, 'BUY'),
+
+	('2020-01-03 00:00:00+00', 2, 'LOGIN'),
+	('2020-01-03 01:00:00+00', 2, 'SEARCH'),
+	('2020-01-03 01:30:00+00', 2, 'BUY'),
+	('2020-01-20 00:00:00+00', 2, 'LOGIN'),
+
+	('2020-02-01 01:00:00+00', 3, 'SEARCH'),
+	('2020-02-14 01:30:00+00', 3, 'BUY'),
+
+	('2020-02-15 00:00:00+00', 4, 'LOGIN'),
+	('2020-02-28 00:00:00+00', 4, 'LOGIN');
+```
+
+### within_interval
+
+Name is terrible, this is just a first test.
+
+Looking for rows where one event occurs within an interval of another, e.g.
+- two LOGIN events within one week
+- BUY event within one hour of SEARCH
+
+#### two LOGIN events within one week
+
+```SQL
+SELECT toolkit_experimental.within_interval(
```

Comment:
One overarching question on this for me is whether we want to use accessors for this or if we want to, for instance, adopt some syntax that constructs a matcher for the the funnel and can be dropped into the where clause like so, I'd also say that many operations on these will probably be best done on some sort of materialization:

```SQL
CREATE MATERIALIZED VIEW funnels as SELECT user, toolkit_experimental.funnel_agg(event, time) funnel
FROM funnel_test
GROUP BY user;

-- Then we can search with the WHERE clause:

SELECT * FROM funnels 

WHERE funnel ? event('LOGIN') -> event('SEARCH', within => '1 hour'::interval) 
OR funnel ? event('LOGIN') -> event('LOGIN', withi n=> '1 hour'::interval, consecutive => true);

```